File: src/ollie/signals/spans.py
"""Map interactions to minimal cold-path span records."""

from __future__ import annotations

from typing import Any

from ollie.primitives import EXTERNAL_INTERACTION, GENERATION


def _attrs_map(ix: dict[str, Any]) -> dict[str, Any]:
    out: dict[str, Any] = {}
    for a in ix.get("attributes") or []:
        if isinstance(a, dict) and a.get("name") is not None:
            out[str(a["name"])] = a.get("value")
    return out


def interaction_status(ix: dict[str, Any]) -> str:
    attrs = _attrs_map(ix)
    if attrs.get("success") is False:
        return "failure"
    status = str(attrs.get("status") or "").strip().lower()
    if status in ("error", "failure", "failed"):
        return "failure"
    return "success"


def interaction_to_span(ix: dict[str, Any]) -> dict[str, Any] | None:
    """Return minimal span for a non-root interaction, or None for root/unknown."""
    if not ix.get("parent_interaction_ref"):
        return None
    prim = str(ix.get("primitive") or "").strip()
    if prim == GENERATION:
        span_type = "llm"
    elif prim == EXTERNAL_INTERACTION:
        span_type = "tool"
    else:
        return None
    name = str(ix.get("name") or span_type)
    status = interaction_status(ix)
    attrs = _attrs_map(ix)
    props: dict[str, Any] = {"kind": span_type, "name": name, "status": status}
    for key in ("latency_ms", "token_count", "finish_reason", "error_type", "model", "provider"):
        # map latency_ms -> duration_ms for warehouse
        if key == "latency_ms" and attrs.get(key) is not None:
            props["duration_ms"] = attrs[key]
        elif attrs.get(key) is not None:
            props[key] = attrs[key]
    inp = str(ix.get("input") or "").strip()
    out = str(ix.get("output") or "").strip()
    span: dict[str, Any] = {
        "type": span_type,
        "name": name,
        "status": status,
        "span_ref": str(ix.get("interaction_ref") or ""),
        "input": {"text": inp} if inp else {},
        "output": {"text": out} if out else {},
        "properties": props,
    }
    if "duration_ms" in props:
        span["duration_ms"] = props["duration_ms"]
    if "token_count" in props:
        span["token_count"] = props["token_count"]
    if "finish_reason" in props:
        span["finish_reason"] = props["finish_reason"]
    parent = ix.get("parent_interaction_ref")
    if parent:
        span["parent_span_ref"] = str(parent)
    return span


def interactions_to_spans(interactions: list[dict[str, Any]]) -> list[dict[str, Any]]:
    spans: list[dict[str, Any]] = []
    for ix in interactions:
        span = interaction_to_span(ix)
        if span and span.get("span_ref"):
            spans.append(span)
    return spans

File: packages/ts/src/signals/spans.ts
import { EXTERNAL_INTERACTION, GENERATION } from "../primitives.js";

export function attrsMap(ix: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const a of (ix.attributes as Record<string, unknown>[]) ?? []) {
    if (a && a.name != null) out[String(a.name)] = a.value;
  }
  return out;
}

export function interactionStatus(ix: Record<string, unknown>): "success" | "failure" {
  const attrs = attrsMap(ix);
  if (attrs.success === false) return "failure";
  const status = String(attrs.status ?? "")
    .trim()
    .toLowerCase();
  if (["error", "failure", "failed"].includes(status)) return "failure";
  return "success";
}

export function interactionToSpan(ix: Record<string, unknown>): Record<string, unknown> | null {
  if (!ix.parent_interaction_ref) return null;
  const prim = String(ix.primitive ?? "").trim();
  let spanType: string | null = null;
  if (prim === GENERATION) spanType = "llm";
  else if (prim === EXTERNAL_INTERACTION) spanType = "tool";
  else return null;

  const name = String(ix.name ?? spanType);
  const status = interactionStatus(ix);
  const attrs = attrsMap(ix);
  const props: Record<string, unknown> = { kind: spanType, name, status };
  if (attrs.latency_ms != null) props.duration_ms = attrs.latency_ms;
  for (const key of ["token_count", "finish_reason", "error_type", "model", "provider"] as const) {
    if (attrs[key] != null) props[key] = attrs[key];
  }
  const inp = String(ix.input ?? "").trim();
  const out = String(ix.output ?? "").trim();
  const span: Record<string, unknown> = {
    type: spanType,
    name,
    status,
    span_ref: String(ix.interaction_ref ?? ""),
    input: inp ? { text: inp } : {},
    output: out ? { text: out } : {},
    properties: props,
  };
  if (props.duration_ms != null) span.duration_ms = props.duration_ms;
  if (props.token_count != null) span.token_count = props.token_count;
  if (props.finish_reason != null) span.finish_reason = props.finish_reason;
  if (ix.parent_interaction_ref) span.parent_span_ref = String(ix.parent_interaction_ref);
  return span;
}

export function interactionsToSpans(interactions: Record<string, unknown>[]): Record<string, unknown>[] {
  const spans: Record<string, unknown>[] = [];
  for (const ix of interactions) {
    const span = interactionToSpan(ix);
    if (span?.span_ref) spans.push(span);
  }
  return spans;
}
